apps/extension/src/storage/migrations/local-v2-migration.test.ts

```diff
@@ -0,0 +1,143 @@
+  let v2ExtStorage: ExtensionStorage<LocalStorageState>;
+  let v3ExtStorage: ExtensionStorage<LocalStorageState>;
+
+  beforeEach(() => {
+    v1ExtStorage = new ExtensionStorage<LocalStorageState>(
+      storageArea,
+      localDefaults,
+      LocalStorageVersion.V1,
+    );
+
+    v2ExtStorage = new ExtensionStorage<LocalStorageState>(
+      storageArea,
+      localDefaults,
+      LocalStorageVersion.V2,
+      {
+        [LocalStorageVersion.V1]: localV1Migrations,
+      },
+      {
+        [LocalStorageVersion.V1]: LocalStorageVersion.V2,
+      },
+    );
+
+    v3ExtStorage = new ExtensionStorage<LocalStorageState>(
+      storageArea,
+      localDefaults,
+      LocalStorageVersion.V3,
+      {
+        [LocalStorageVersion.V1]: localV1Migrations,
+        [LocalStorageVersion.V2]: localV2Migrations,
+      },
+      {
+        [LocalStorageVersion.V1]: LocalStorageVersion.V2,
+        [LocalStorageVersion.V2]: LocalStorageVersion.V3,
+      },
+    );
+  });
+
+  test('non-affected fields stay the same', async () => {
+    await v1ExtStorage.set('fullSyncHeight', 9483729);
+    const syncHeight = await v3ExtStorage.get('fullSyncHeight');
+    expect(syncHeight).toEqual(9483729);
+  });
+
+  describe('frontends', () => {
+    test('not set frontend gets ignored', async () => {
+      await v1ExtStorage.set('frontendUrl', '');
+      const url = await v3ExtStorage.get('frontendUrl');
+      expect(url).toEqual('');
+    });
+
+    test('have no change if user already selected frontend in registry', async () => {
+      const registryClient = new ChainRegistryClient();
+      const { frontends } = registryClient.bundled.globals();
+      const suggestedFrontend = sample(frontends.map(f => f.url));
+      await v1ExtStorage.set('frontendUrl', suggestedFrontend);
+      const url = await v3ExtStorage.get('frontendUrl');
+      expect(url).toEqual(suggestedFrontend);
+    });
+
+    test('user gets migrated to suggested frontend', async () => {
+      const registryClient = new ChainRegistryClient();
+      const { frontends } = registryClient.bundled.globals();
+      await v1ExtStorage.set('frontendUrl', 'http://badfrontend.void');
+      const url = await v3ExtStorage.get('frontendUrl');
+      expect(url).not.toEqual('http://badfrontend.void');
+      expect(frontends.map(f => f.url).includes(url!)).toBeTruthy();
+    });
+
+    test('works from v2 storage as well', async () => {
+      const registryClient = new ChainRegistryClient();
+      const { frontends } = registryClient.bundled.globals();
+      await v2ExtStorage.set('frontendUrl', 'http://badfrontend.void');
+      const url = await v3ExtStorage.get('frontendUrl');
+      expect(url).not.toEqual('http://badfrontend.void');
+      expect(frontends.map(f => f.url).includes(url!)).toBeTruthy();
+    });
+  });
+
+  describe('grpcEndpoint', () => {
+    test('not set gets ignored', async () => {
+      await v1ExtStorage.set('grpcEndpoint', undefined);
```

Comment:
what happens in scenarios where the `grpcEndpoint` or `frontendUrl` are _null_ rather than undefined?

Reply:
We don't set those as null, but null fields are possible (though, still of the opinion null is a programming language error)